updateable count labels

gogonzo opened the pull request:
closes #173 #174 #167 
Tested on: chrome, safari and firefox with default shiny, BS3-5. No major styling/js issues. 

In this PR we've changed the way how the labels are created and updated. Basically, counts are the separate shiny/html elements which have their own `id`, which they are identified by. 
1. We introduced new shiny inputs to create and update values of the labels:
- countBar - creates n ui elements for each choice with progressbar and text label
- countBar - creates a ui html element with progressbar and text label
- make_count_text - composes the label `$choice ($filtered/$unfiltered)`

2. We've created respective update* methods to keep the labels up to date
- updateCountBars
- updateCountBar
- updateCountText

3. To update the counts without triggering reactive cycle we created two js handlers which change a bar and a label found using `id`.

4. Instead of `observeEvent(filtered_counts(), ...)` we used `renderUI(... filtered_counts())` which serves as the observer for the visible item only!

5. In the ChoicesFilterState we applied [this](https://github.com/dreamRs/shinyWidgets/issues/392) to trigger change of the selection only when the picker is closed - closing #167 

![Untitled Diagram drawio (2)](https://user-images.githubusercontent.com/6959016/219326830-aa6d4fa0-51d3-493b-b65c-5bda5328d6c4.png)


You can find examples in #165 and one can use also `teal.gallery::launch_app`.

asbates commented:
I would consider having the progress bars CSS as the following.

- Add border radius to right side of unfiltered bar to match the left side
- Use variables where possible to maintain user theme if they have one set. (`var(--color, othercolor)` will use `--color` if available, otherwise `othercolor`.
- Instead of built in css lightblue, use Shiny's default primary color to better match theme.

```css
.progress-bar.state-count-bar-filtered {
  background-color: var(--bs-primary, rgb(51, 122, 183));
  opacity: 0.8;
  color: var(--bs-body-color, var(--dark, #333333));
  overflow: visible;
  text-align: left;
  white-space: nowrap;
}

.progress-bar.state-count-bar-unfiltered {
  background-color: var(--bs-primary, rgb(51, 122, 183));
  opacity: 0.439;
  border-top-right-radius: var(--bs-border-radius, .375rem);
  border-bottom-right-radius: var(--bs-border-radius, .375rem);
}
```

donyunardi commented:
Thanks for working on this @gogonzo. I think this is looking great already. The js handler solution works great on addressing our reactive cycle issue.

@lcd2yyz 
Could you please look at this PR, specifically the progress bar?

gogonzo commented on inst/css/filter-panel.css:
```diff
@@ -251,3 +251,22 @@ a.remove_all:hover {
+  margin: 0px;
+}
+
+.progress-bar.state-count-bar-filtered {
+  background-color: lightblue !important;
+  color: var(--bs-body-color, var(--dark, #333333));
+  overflow: visible;
+  text-align: left;
+  white-space: nowrap;
+}
+
+.progress-bar.state-count-bar-unfiltered {
+  background-color: rgba(173, 216, 230, 0.439) !important;
+}
```
@asbates do you know how to apply alpha/opacity to the background only? Opacity affects text and background in the same time which means that it could be barely visible,

![image](https://user-images.githubusercontent.com/6959016/219609727-d394a1ee-bd62-4689-985b-0cbc88d1f4da.png)


```suggestion
.progress-bar.state-count-bar-filtered {
  background-color: var(--bs-primary, rgb(51, 122, 183));
  opacity: 0.8;
  color: var(--bs-body-color, var(--dark, #333333));
  overflow: visible;
  text-align: left;
  white-space: nowrap;
}

.progress-bar.state-count-bar-unfiltered {
  background-color: var(--bs-primary, rgb(51, 122, 183));
  opacity: 0.439;
  border-top-right-radius: var(--bs-border-radius, .375rem);
  border-bottom-right-radius: var(--bs-border-radius, .375rem);
}
```

asbates replied:
I do not. Bootstrap uses hex for color variables and as of now there isn't a way to add alpha to an existing hex color. (apparently this is in the works). So setting opacity on the whole bars is probably not the way to go. The text looked light to me but I wanted to bring it up just in case it was just me.

mhallal1 commented on R/count_labels.R:
```diff
@@ -0,0 +1,213 @@
+#' Progress bars with labels
+#'
+#' `shiny` element showing progressbar counts. Each element can has
```
```suggestion
#' `shiny` element showing progressbar counts. Each element can have a
```

mhallal1 commented on R/count_labels.R:
```diff
@@ -0,0 +1,213 @@
+#' )
+#'
+#'
+#' shinyApp(
```
![Screenshot from 2023-02-20 11-11-47](https://user-images.githubusercontent.com/86970066/220077315-459f1b9c-a945-44b7-952b-83d8e5e99ace.png)

mhallal1 replied:
fixed, thanks

mhallal1 commented on R/count_labels.R:
```diff
@@ -0,0 +1,213 @@
+
+#' Make a count text
+#'
+#' Returns a text describing filtered counts. Text is composed in following way:
```
```suggestion
#' Returns a text describing filtered counts. The text is composed in the following way:
```

mhallal1 reviewed (commented):
Issue created while reviewing this: https://github.com/insightsengineering/teal.slice/issues/203

mhallal1 commented on R/count_labels.R:
```diff
@@ -0,0 +1,214 @@
+#'  Length should be the same as `choices`.
+#' @param countsnow (`numeric`) actual counts of each element.
+#'  Length should be the same as `choices`.
+#' @param counttotal (`numeric(1)`) Determines the size of the whole progress bar.
+#'  For example, in case there are three choices with some counts, one might like
+#'  to display each progress bar with the relative size to the `totalcount`.
+#' @return list of `shiny.tag`
```
`counttotal` is an an argument for  `countBars`

mhallal1 replied:
neither the session

gogonzo replied:
session is there - countBars and updateCountBars have a shared documentation page

![image](https://user-images.githubusercontent.com/6959016/220140007-f1f89b33-9683-4030-8e33-a154a4b7177f.png)

mhallal1 replied:
fair enough,

mhallal1 reviewed (approved):
Tested locally and on BEE with multiple apps and conditions.
Works as expected.

chlebowa commented on R/utils.R:
```diff
@@ -104,3 +104,22 @@ get_teal_bs_theme <- function() {
+#' @return HTML code that includes `JS` files
+#' @keywords internal
+include_js_files <- function(pattern = NULL, except = NULL) {
+  checkmate::assert_character(except, min.len = 1, any.missing = FALSE, null.ok = TRUE)
```
`min.len = 1` precludes NULL

gogonzo replied:
I've removed exclude at all as it's not used in this package. However, `min.len = 1` and `null.ok = TRUE` can be specified together which prevents someone from specifying `character(0)`

chlebowa replied:
But if `min.len = 1`, NULL will not pass as it has length 0. `min.len = 1` is sufficient here.

gogonzo replied:
```r
checkmate::assert_character(NULL, min.len = 1, any.missing = FALSE, null.ok = TRUE)
#> silent
```

chlebowa replied:
I stand corrected.

chlebowa commented on R/count_labels.R:
```diff
@@ -0,0 +1,211 @@
+#' )
+#' }
+#' @keywords internal
+countBars <- function(inputId, choices, countsmax, countsnow = NULL) {
+  checkmate::assert_string(inputId)
+  checkmate::assert_vector(choices)
+  checkmate::assert_numeric(countsmax, len = length(choices))
+  checkmate::assert_numeric(countsnow, len = length(choices), null.ok = TRUE)
+  if (!is.null(countsnow)) {
+    checkmate::assert_true(all(countsnow <= countsmax))
+  }
+
+  ns <- NS(inputId)
+  counttotal <- sum(countsmax)
+
+  lapply(seq_along(choices), function(i) {
+    choice <- as.character(choices[i])
+    countmax <- countsmax[i]
+    countnow <- if (is.null(countsnow)) 0 else countsnow[i]
+
+    countBar(
+      inputId = ns(i),
+      label = choice,
+      countmax = countmax,
+      countnow = countnow,
+      counttotal = counttotal
+    )
+  })
+}
```
I suggest `mapply`:

```suggestion
countBars <- function(inputId, choices, countsmax, countsnow = NULL) {
  checkmate::assert_string(inputId)
  checkmate::assert_vector(choices)
  checkmate::assert_numeric(countsmax, len = length(choices))
  checkmate::assert_numeric(countsnow, len = length(choices), null.ok = TRUE)
  if (!is.null(countsnow)) {
    checkmate::assert_true(all(countsnow <= countsmax))
  } else {
    countsnow <- rep(0L, length(choices))
  }

  ns <- NS(inputId)

  mapply(
    countBar,
    inputId = ns(seq_along(choices)),
    label = as.character(choices),
    countmax = countsmax,
    countnow = countsnow,
    MoreArgs = list(
      counttotal = sum(countsmax)
    ),
    SIMPLIFY = FALSE, USE.NAMES = FALSE)
}
```

chlebowa commented on R/count_labels.R:
```diff
@@ -0,0 +1,211 @@
+}
+
+#' @rdname countBars
+updateCountBars <-  function(session = getDefaultReactiveDomain(), inputId, choices,
+                                  countsmax, countsnow = NULL) {
+    checkmate::assert_string(inputId)
+    checkmate::assert_vector(choices)
+    checkmate::assert_numeric(countsmax, len = length(choices))
+    checkmate::assert_numeric(countsnow, len = length(choices), null.ok = TRUE)
+
+    ns <- NS(inputId)
+    counttotal <- sum(countsmax)
+    lapply(seq_along(choices), function(i) {
+      choice <- choices[i]
+      countmax <- countsmax[i]
+      countnow <- if (is.null(countsnow)) countmax else countsnow[i]
+      updateCountBar(
+        inputId = ns(i),
+        label = choice,
+        countmax = countmax,
+        countnow = countnow,
+        counttotal = counttotal
+      )
+    })
+  invisible(NULL)
+}
```
Also suggest `mapply` here

```suggestion
updateCountBars <-  function(session = getDefaultReactiveDomain(), inputId, choices,
                                  countsmax, countsnow = NULL) {
    checkmate::assert_string(inputId)
    checkmate::assert_vector(choices)
    checkmate::assert_numeric(countsmax, len = length(choices))
    checkmate::assert_numeric(countsnow, len = length(choices), null.ok = TRUE)
    if (!is.null(countsnow)) {
      checkmate::assert_true(all(countsnow <= countsmax))
    } else {
      countsnow <- rep(0L, length(choices))
    }

    ns <- NS(inputId)
    mapply(
      updateCountBar,
      inputId = ns(seq_along(choices)),
      label = choices,
      countmax = countsmax,
      countnow = countsnow,
      MoreArgs = list(
        counttotal = sum(countsmax)
      )
    )
  invisible(NULL)
}
```

gogonzo commented:
After addressing all comments from @chlebowa and @BLAZEWIM I'm using @mhallal1 approval. If you still see the room for improvement please create an issue 👍
